### DISCUSSION TITLE
Directly Manipulate position

### DISCUSSION THREAD

**Discussion by blacksino** [InteractiveComputerGraphics/PositionBasedDynamics#117] [Q&A]
Hi,I am a new to PositionBaseDynamics,I am tring to simulate defomation of human organs by PBD,I was wondering if there is a possible solution to manipulate specific vertex's positions in tetrahedron mesh to make the mesh gradually deform to target shape by using FEMTetraConstraint and pre-define shape constraint,if so, how to implement such pre-define shape constraint？
For example,If I want to transform a sphere into an ellipsoid gradually , should I sample some points uniformly on the sphere, and then give the target positions of these sampling points, and then project the FEM tetrahedron constraints and the target position constraints of the sampling points separately？

**Comment by janbender** [comment]
I did not get your problem. Do you want to do some kind of morphing?

**Reply by blacksino** [reply]
Exactly, what I am trying to implement is a kind of morphing using several known points, Just like ParticleSfT [did.](http://igt.ip.uca.fr/encov/publications/pubfiles/2016_Ozgur_etal_IJCV_sft.pdf)
Such as applying drap and drop on mesh, with knowing target position, I want to know how the mesh deforms.
For another example, if I know the positions of some key points on the mesh after deformation, I want to infer how the mesh deforms through the positions of these key points, although this is an under-constrained problem.

**Comment by janbender** [comment]
The FEMTetraConstraint simulates elasticity after an object has deformed but it cannot be used to pull vertices to target positions. You could animate the vertices by directly setting their positions and then go in the solver loop with the FEMTetraConstraint. I assume that only some of the vertices are set to target positions.

**Reply by blacksino** [reply]
Yeah, That's the way I have tried, However, there are several problems:
1. If I directly set position and then go in the solver loop with the FEMTetraConstraint, the Gauss-seidel solver seems to be quite unstable, in each sub-step, it needs many iterations to converge, especially when it came to large deformation or large displacement of control vertex, correspondingly, if I reduce the times of GS-iteration( corresponding variable in code: **m_maxIterations**), the deformation is unsatisfied, for example : collapse, distortion( maybe caused by not setting self collisions constraint for deformables? I think), etc.
2. If I set position step by step with an increase number of GS-iteration, it's getting better(not that much better, still got collapse sometimes) yet much slower.

I was wondering if there is something I can do to make the process of deformation stable while reducing the number of iterations,i.e, replace GS-solver with some kind of parallelizable non-linear solver, or reduce the number of vertex of mesh?

**Reply by blacksino** [reply]
Another theoretical confusion: the GS-type iteration solver mentioned in the paper and the one implemented in this repository seem to be irrelevant with the GS iteration algorithm for linear equation, the solver in PBD  iteratively solves all the constraint equations serially, there is no guarantee of convergence of this method, for overdetermined equations, the least-squares solution may not be obtained, and oscillations may occur. Is there any method to avoid the situation mentioned above？

Sorry for that much confusion. There may be many mistakes in my understanding of the paper and source code, but I am really interested in PBD, hope you could help me out, thank you in advance. :smile:

**Reply by janbender** [reply]
No, there is no guarantee if you have an overdetermined system. A GS solver only converges if your system is symmetric positive definite.

**Reply by blacksino** [reply]
Thanks, well, it does make sense that there is no guarantee for an overdetermined system.
The problem that I have been struggling with is some kind of Coutour Constraint, I want the tetrahedron model to deform to a specific shape that fits a 2D-contour, which means, the silhouette of the tetrahedron from a specific perspective must fit a pre-define curve.

However, I found that it is extremely difficult to develop such kind of constraint, particles on the silhouette update every time the model deform.
Really appreciate it if you can give us some advice.
![Peek 2022-02-11 16-37](https://user-images.githubusercontent.com/44363764/153560538-be26ffd9-e714-409b-8805-5b796fa30df9.gif)

**Reply by blacksino** [reply]
By the way, "particles on the silhouette" update means every time the model deforms, the algorithm extracts the silhouette of the tetrahedron　again, so silhouette updates, particles on the silhouette updates too.

**Comment by janbender** [comment]
It might be difficult for the solver if several positions are fixed and cannot be moved. Maybe you could try not to move the positions directly but to pull them to the target position by a distance constraint.
